### PR TITLE
include: improve compatibility with C++ apps.

### PR DESCRIPTION
**Pull request by nzmichaelh**
This patch lets a C++ application use more of Zephyr by adding guards
and changeing some constructs to the C++11 equivalent.

Changes include:

- Adding guards
- Switching to static_assert
- Switching to a template for ARRAY_SIZE as g++ doesn't have the
  builtin.
- Re-ordering designated initialisers to match the struct field order
  as G++ only supports simple designated initialisers.

Signed-off-by: Michael Hope <mlhx@google.com>

### CODE REVIEW

**Comment by nzmichaelh** [comment]
This is failing as the tree is unstable:

zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -Wl,--end-group -L"/opt/sdk/zephyr-sdk-0.9.2/sysroots/armv5-zephyr-eabi/usr/lib/arm-zephyr-eabi/6.2.0/armv7e-m" -L/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/nrf52840_pca10056/samples/subsys/usb/cdc_acm/test_x86/zephyr -lgcc -Wl,--print-memory-usage -nostartfiles -nodefaultlibs -nostdlib -static -no-pie -Wl,--fatal-warnings -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none && :
Memory region         Used Size  Region Size  %age Used
           FLASH:       67364 B         1 MB      6.42%
            SRAM:       16756 B       256 KB      6.39%
        IDT_LIST:         164 B         2 KB      8.01zephyr/libzephyr.a(usb_device.c.obj): In function `usb_transfer_work':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/usb/usb_device.c:1109: undefined reference to `usb_dc_ep_mps'

**Comment by nzmichaelh** [comment]
Tree is now stable, yay!

File: ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h [hunk 1]
```diff
@@ -58,6 +58,10 @@
 #ifndef __UECC_PLATFORM_SPECIFIC_H_
 #define __UECC_PLATFORM_SPECIFIC_H_
 
+#ifdef __cplusplus
```
Comment (nashif) [inline ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h]: please split this into its own commit and send tinycrypt patches to https://github.com/01org/tinycrypt/pulls

File: ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h [hunk 2]
```diff
@@ -78,4 +82,8 @@
 
 int default_CSPRNG(uint8_t *dest, unsigned int size);
 
+#ifdef __cplusplus
```
Comment (nashif) [inline ext/lib/crypto/tinycrypt/include/tinycrypt/ecc_platform_specific.h]: ditto

**Review by nashif — CHANGES_REQUESTED** [review]

**Review by tyrantlizard77 — DISMISSED** [review]
LGTM but please split out and upstream the bits that touch tinycrypt.

**Comment by nzmichaelh** [comment]
@andrewboie @nashif  I've split out the ext/ changes.

I also went through the most common driver files and added guards, and added a mainly compile related test to see that they stay compatible in the future.

File: tests/application_development/cpp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
+
+FILE(GLOB app_sources src/*.cpp)
```
Comment (mbolivar) [inline tests/application_development/cpp/CMakeLists.txt]: I think @SebastianBoe prefers not using GLOB in Zephyr's CMake files.

**Review by mbolivar — COMMENTED** [review]
led_strip changes are fine. One potential issue on the test CMakeLists.txt.

File: tests/application_development/cpp/CMakeLists.txt (hunk 3 above)
Comment (nashif) [reply tests/application_development/cpp/CMakeLists.txt]: git grep "GLOB app_sources" | wc -l
261

**Review by nashif — APPROVED** [review]

File: tests/application_development/cpp/CMakeLists.txt (hunk 3 above)
Comment (mbolivar) [reply tests/application_development/cpp/CMakeLists.txt]: Cool, never mind :+1:

**Comment by nzmichaelh** [comment]
I've updated spi_legacy.h as well to fix the compile test.

**Comment by nzmichaelh** [comment]
@nashif  Hi Anas.  Could you install lib32stdc++-7-dev on the Shippable configuration?  It's needed for C++ apps to link in the POSIX m32 config.

**Comment by nzmichaelh** [comment]
PR sent for adding the m32 version of libsdc++ to the Shippable config.

**Comment by nashif** [comment]
recheck

File: tests/application_development/cpp/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+static struct foo foos[5];
+BUILD_ASSERT_MSG(ARRAY_SIZE(foos) == 5, "expected 5 elements");
+
+static int test_init(struct device *dev)
+{
+	return 0;
+}
+
+SYS_INIT(test_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
Comment (nashif) [inline tests/application_development/cpp/src/main.cpp]: why are you starting the test function this way and not via test_main?

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nzmichaelh** [comment]
recheck

File: tests/application_development/cpp/src/main.cpp (hunk 4 above)
Comment (nzmichaelh) [reply tests/application_development/cpp/src/main.cpp]: Clarified that this is to test SYS_INIT itself.

**Comment by nashif** [comment]
recheck

File: tests/application_development/cpp/testcase.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+tests:
+  test:
+    tags: cpp
```
Comment (nashif) [inline tests/application_development/cpp/testcase.yaml]: if this is going to be a compile test only, i.e. this test is not supposed to run, then please add:

build_only: true

Also, AFAIK c++ is still not supported with native_posix, so you want to add arch_exclude: posix as well.

**Comment by nashif** [comment]
@nzmichaelh  please rebase

**Comment by nzmichaelh** [comment]
@nashif I've rebased and added the flags to testcase.yaml.  Let's see how the build goes...

**Comment by nashif** [comment]
@nzmichaelh 
> @nashif I've rebased and added the flags to testcase.yaml. Let's see how the build goes...

thanks :)

**Review by nashif — APPROVED** [review]
